### PR TITLE
Base unit multiplier sdk adjustments

### PR DESCRIPTION
**Pull request by throwbackjams**
Related to james/ell-12-update-base-multiplier-to-support-long. Fixes ELL-12

Update rust sdk unit methods to account for the raw base unit to Phoenix BaseUnit adjustment factor; Change method names to be clearer and less ambiguous. 

Next step: update typescript sdk accordingly

### CODE REVIEW

File: rust/Cargo.toml [hunk 1]
```diff
@@ -20,7 +20,7 @@ anyhow = "1.0.52"
 solana-sdk = "1.10.32"
 borsh = "0.9.3"
 rand = "0.7.3"
-phoenix-types = { git = "https://github.com/Ellipsis-Labs/phoenix-types", branch = "master" }
+phoenix-types = { path = "../../phoenix-types" }
```
Comment (eugene-chen) [inline rust/Cargo.toml:23]: note to fix this before merging

Comment (throwbackjams) [reply rust/Cargo.toml]: Ah thank you. Reverted in the latest commit

Comment (eugene-chen) [reply rust/Cargo.toml]: no worries. i stick this in basically all my PRs and leave comments like this (https://github.com/Ellipsis-Labs/phoenix/pull/249#discussion_r1069911937) just to make sure it gets reverted before merging. ideally we would have CI set up to yell at you before merging so we don't need these manual comments

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 2]
```diff
@@ -101,24 +103,33 @@ impl SDKClientCore {
     /// Converts base units to base lots. For example if the base currency was a Widget and you wanted to
     /// convert 3 Widgets to base lots you would call sdk.base_unit_to_base_lots(3.0). This would return
     /// the number of base lots that would be equivalent to 3 Widgets.
-    pub fn base_units_to_base_lots(&self, base_units: f64) -> u64 {
-        (base_units * self.base_multiplier as f64 / self.base_lot_size as f64) as u64
+    pub fn raw_base_units_to_base_lots(&self, raw_base_units: f64) -> u64 {
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:106]: fix doc

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 3]
```diff
@@ -101,24 +103,33 @@ impl SDKClientCore {
...
     /// RECOMMENDED:
-    /// Converts base amount to base lots. For example if the base currency was a Widget with 9 decimals and you wanted to
+    /// Converts base atoms to base lots. For example if the base currency was a Widget with 9 decimals, where 1 atom is 1e-9 of one Widget and you wanted to
     /// convert 3 Widgets to base lots you would call sdk.base_amount_to_base_lots(3_000_000_000). This would return
-    /// the number of base lots that would be equivalent to 3 Widgets.
-    pub fn base_amount_to_base_lots(&self, base_amount: u64) -> u64 {
-        base_amount / self.base_lot_size
+    /// the number of base lots that would be equivalent to 3 Widgets or 3 * 1e9 Widget atoms.
+    pub fn base_atoms_to_base_lots(&self, base_atoms: u64) -> u64 {
+        base_atoms / self.base_lot_size //Lot size is the number of atoms in a lot
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:124]: nit: space after // (and the same below)

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 4]
```diff
@@ -130,43 +141,43 @@ impl SDKClientCore {
     }
 
     /// RECOMMENDED:
-    /// Converts quote amount to quote lots. For example if the quote currency was USDC with 6 decimals and you wanted to
-    /// convert 3 USDC to quote lots you would call sdk.quote_amount_to_quote_lots(3_000_000). This would return
-    /// the number of quote lots that would be equivalent to 3 USDC.
-    pub fn quote_amount_to_quote_lots(&self, quote_amount: u64) -> u64 {
-        quote_amount / self.quote_lot_size
+    /// Converts quote atoms to quote lots. For example if the quote currency was USDC with 6 decimals and you wanted to
+    /// convert 3 USDC, or 3_000_000 USDC atoms, to quote lots you would call sdk.quote_amount_to_quote_lots(3_000_000). This would return
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:145]: nit: fix doc `quote_amount_to_quote_lots`

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 5]
```diff
@@ -130,43 +141,43 @@ impl SDKClientCore {
...
     }
 
-    /// Takes in a quote amount and prints it as a human readable string to the console
+    /// Takes in a number of quote atoms, converts to floating point number of whole tokens, and prints it as a human readable string to the console
     pub fn print_quote_amount(&self, quote_amount: u64) {
         println!("{}", get_decimal_string(quote_amount, self.quote_decimals));
     }
 
-    /// Takes in a base amount and prints it as a human readable string to the console
+    /// Takes in a number of quote atoms, converts to floating point number of whole tokens, and prints it as a human readable string to the console
```
Comment (eugene-chen) [inline rust/phoenix-sdk-core/src/sdk_client_core.rs:180]: base

File: rust/phoenix-sdk/src/sdk_client.rs [hunk 6]
```diff
@@ -286,6 +286,7 @@ impl SDKClient {
         let tick_size_in_quote_atoms_per_base_unit =
             header.get_tick_size_in_quote_atoms_per_base_unit().into();
         let num_base_lots_per_base_unit = market.get_base_lots_per_base_unit().into();
+        let raw_base_units_to_base_units = header.raw_base_units_per_base_unit.max(1);
```
Comment (eugene-chen) [inline rust/phoenix-sdk/src/sdk_client.rs:289]: why `max`? shouldn't this always be set? or is this to support old markets that don't have this field set (and so currently would be read as 0)?

Comment (eugene-chen) [reply rust/phoenix-sdk/src/sdk_client.rs:289]: can we just redeploy and remove?

Comment (jarry-xiao) [reply rust/phoenix-sdk/src/sdk_client.rs:289]: We can, i don't really want to redeploy the devnet market though. We can probably add something really hacky where in new order we update the header if this field is 0 and then revert the change

File: rust/phoenix-sdk-core/src/sdk_client_core.rs [hunk 7]
```diff
@@ -78,8 +78,10 @@ pub struct MarketMetadata {
     pub base_lot_size: u64,
     pub tick_size_in_quote_atoms_per_base_unit: u64,
     pub num_base_lots_per_base_unit: u64,
+    /// The adjustment factor to convert from the raw base unit (i.e. 1 BONK token) to the Phoenix BaseUnit (which may be a multiple of whole tokens).
+    /// The adjustment factor is almost always 1, unless one base token is worth less than one quote atom (i.e. 1e-6 USDC)
```
[the remark this answers is not in the document]
Comment (throwbackjams) [reply rust/phoenix-sdk-core/src/sdk_client_core.rs:82]: Good catch. Changed to reflect in latest commit

File: rust/phoenix-sdk-core/src/sdk_client_core.rs (hunk 2 above)
Comment (throwbackjams) [reply rust/phoenix-sdk-core/src/sdk_client_core.rs:106]: I might just be missing it but what is wrong here? Perhaps "Converts raw base units" ?

File: rust/phoenix-sdk-core/src/sdk_client_core.rs (hunk 3 above)
Comment (throwbackjams) [reply rust/phoenix-sdk-core/src/sdk_client_core.rs]: Changed to reflect in latest commit

File: rust/phoenix-sdk-core/src/sdk_client_core.rs (hunk 5 above)
Comment (throwbackjams) [reply rust/phoenix-sdk-core/src/sdk_client_core.rs]: Changed to reflect in latest commit

File: rust/phoenix-sdk-core/src/sdk_client_core.rs (hunk 4 above)
Comment (throwbackjams) [reply rust/phoenix-sdk-core/src/sdk_client_core.rs:145]: I believe `quote_amount_to_quote_lots` is deleted and the docs now match with the latest function name `quote_units_to_quote_lots`

File: rust/phoenix-sdk-core/src/sdk_client_core.rs (hunk 2 above)
Comment (eugene-chen) [reply rust/phoenix-sdk-core/src/sdk_client_core.rs:106]: yes, and the function name in the doc doesn't match the actual function name

File: rust/phoenix-sdk-core/src/sdk_client_core.rs (hunk 4 above)
Comment (eugene-chen) [reply rust/phoenix-sdk-core/src/sdk_client_core.rs:145]: The doc says `quote_amount_to_quote_lots` still?

File: rust/phoenix-sdk-core/src/sdk_client_core.rs (hunk 2 above)
Comment (throwbackjams) [reply rust/phoenix-sdk-core/src/sdk_client_core.rs:106]: Ah gotcha thanks! Changed in the latest commit

File: rust/phoenix-sdk-core/src/sdk_client_core.rs (hunk 4 above)
Comment (throwbackjams) [reply rust/phoenix-sdk-core/src/sdk_client_core.rs]: Ah I see thanks! Changed in the latest commit

**Review by eugene-chen — APPROVED** [review]
LGTM. Please merge https://github.com/Ellipsis-Labs/phoenix-types/pull/17 first, then update the Cargo.toml here and make sure it still compiles

**Pull request merged**